cpp/src/merge/merge.cu

```diff
@@ -0,0 +1,481 @@
+      }
+
+    //collect index columns for lhs, rhs, resp.
+    //
+    std::vector<cudf::column_view> left_index_cols;
+    std::vector<cudf::column_view> right_index_cols;
+    bool nullable{false};
+
+    //TODO: use `table_view select(std::vector<cudf::size_type> const& column_indices) const;`
+    //(when it becomes available; currently in PR# 3144)
+    //
+    for(auto&& indx: key_cols)
+      {
+        const cudf::column_view& left_col = left_table.column(indx);
+        const cudf::column_view& right_col= right_table.column(indx);
+
+        //for the purpose of generating merged indices, there's
+        //no point looking into _all_ table columns for nulls,
+        //just the index ones:
+        //
+        if( left_col.has_nulls() || right_col.has_nulls() )
+          nullable = true;
+        
+        left_index_cols.push_back(left_col);
+        right_index_cols.push_back(right_col);
+      }
+    cudf::table_view index_left_view{left_index_cols};   //table_view move cnstr. would be nice
+    cudf::table_view index_right_view{right_index_cols}; //same...
+
+    //extract merged row order according to indices:
+    //
+    rmm::device_vector<index_type>
+      merged_indices = generate_merged_indices(index_left_view,
+                                               index_right_view,
+                                               column_order,
+                                               null_precedence,
+                                               nullable);
+
+    //create merged table:
+    //
+    std::vector<std::unique_ptr<column>> v_merged_cols;
+    v_merged_cols.reserve(n_cols);
+
+    ColumnMerger merger{merged_indices};
+    
+    for(auto i=0;i<n_cols;++i)
+      {
+        const auto& left_col = left_table.column(i);
+        const auto& right_col= right_table.column(i);
+
+        //not clear yet what must be done for STRING:
+        //
+        //if( left_col.type().id() != STRING )
+        //  continue;//?
+
+        auto merged = cudf::experimental::type_dispatcher(left_col.type(),
+                                                          merger,
+                                                          left_col,
+                                                          right_col);
+        v_merged_cols.emplace_back(std::move(merged));
+      }
+    
+    return std::make_unique<cudf::experimental::table>(std::move(v_merged_cols));
+}
+
+}  // namespace detail
+
+std::unique_ptr<cudf::experimental::table> merge(table_view const& left_table,
+                                                 table_view const& right_table,
+                                                 std::vector<cudf::size_type> const& key_cols,
+                                                 std::vector<cudf::order> const& column_order,
+                                                 std::vector<cudf::null_order> const& null_precedence){
```

Comment:
Missing `cudaStream_t` and `rmm::mr::device_memory_resource` parameters.

Reply:
It's a bit late in the process to change the API. Are we sure we want to modify it to add these parameters? 

Also, should `null_precedence = {}` default to empty vector?

Reply:
Done.